File: problem46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # Base case
        if len(nums) == 1:
            return [[nums[0]]]

        # It is backtracking: returning all possible combinations
        permut = [] # this is possible permutations
        used = set()  # this used for know which interger is used since all integreg are differnt
        def backtrack(used, path):
            # when path includes all 
            if len(path) == len(nums): # all know all path has equal number of nums
                print(permut)
                permut.append(path[:])

            for i in range(len(nums)):
                if nums[i] in used:
                    continue
                # if num is not used, add to used
                used.add(nums[i])
                path.append(nums[i]) # then expand path by num 
                backtrack(used, path) # call function again for next num
                used.remove(nums[i]) # after it has all numbers, return last element from used
                path.remove(nums[i]) # also remove last elemt from the path too see new path

        backtrack(used, [])
        return permut

File: problem46.cpp
#include <vector>
#include <unordered_set>
#include <iostream>
using namespace std;
class Solution
{
public:
    vector<vector<int>> permutation;
    vector<int> current;
    unordered_set<int> used;
    void backtrack(vector<int> &nums)
    {
        // if the path has all element push to permutation
        if (current.size() == nums.size())
        {
            permutation.push_back(current); // the current path
            return;
        }
        for (int i = 0; i < nums.size(); i++)
        {
            if (used.contains(nums[i]))
            {
                continue;
            }
            used.insert(nums[i]);       // add to used set
            current.push_back(nums[i]); // add to current combination
            backtrack(nums);            // call it again
            used.erase(nums[i]);
            current.pop_back();
        }
    }

    vector<vector<int>> permute(vector<int> &nums)
    {
        permutation.clear();
        current.clear();
        backtrack(nums);
        return permutation;
    }
};